### PR TITLE
client/mds: revoke the Fx caps to trigger dirty metadatas to be flushed from client

### PR DESCRIPTION
**Pull request by lxbsz**
If the 'Fx' caps is issued the client will cache the dirty metadatas,
while if a write request, mkdir request for example, comes it will
try to update the parents' mtime without trigger to revoke the
corresponding caps. This will cause the mtime will be overwrote
later after the dirty caps is flushed from clients.

We need to acquire the xlock for the parent inode.

And also we could voluntarily drop Xx/Lx/Ax caps for some requests
to avoid revocation msgs.

Fixes: https://tracker.ceph.com/issues/61584

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by vshankar** [comment]
jenkins test api

**Comment by vshankar** [comment]
jenkins test windows

**Comment by vshankar** [comment]
@lxbsz Thanks for pointing to this. I've noted it my backlog and will to it on Monday (4th Sep).

**Comment by lxbsz** [comment]
> @lxbsz Thanks for pointing to this. I've noted it my backlog and will to it on Monday (4th Sep).

Sure, thanks @vshankar

**Comment by vshankar** [comment]
Hey Xiubo, I left a note on the tracker - https://tracker.ceph.com/issues/61584#note-5

**Comment by lxbsz** [comment]
> Hey Xiubo, I left a note on the tracker - https://tracker.ceph.com/issues/61584#note-5

Yeah, I replied it. Thanks @vshankar

File: src/mds/MDCache.cc [hunk 1]
```diff
@@ -2153,10 +2153,9 @@ void MDCache::predirty_journal_parents(MutationRef mut, EMetaBlob *blob,
     pf->version = parent->pre_dirty();
 
     if (do_parent_mtime || linkunlink) {
-      ceph_assert(mut->is_wrlocked(&pin->filelock));
       ceph_assert(mut->is_wrlocked(&pin->nestlock));
       ceph_assert(cfollows == CEPH_NOSNAP);
-      
+
```
Comment (vshankar) [inline src/mds/MDCache.cc:2158]: ditto

File: src/mds/MDCache.cc [hunk 2]
```diff
@@ -2142,7 +2142,7 @@ void MDCache::predirty_journal_parents(MutationRef mut, EMetaBlob *blob,
   while (parent) {
     //assert(cur->is_auth() || !primary_dn);  // this breaks the rename auth twiddle hack
     ceph_assert(parent->is_auth());
-    
+
```
Comment (vshankar) [inline src/mds/MDCache.cc:2145]: unrelated change, please drop.

**Comment by vshankar** [comment]
> If the 'Fx' caps is issued the client will cache the dirty metadatas, while if a write request, mkdir request for example, comes it will try to update the parents' mtime without trigger to revoke the corresponding caps. This will cause the mtime will be overwrote later after the dirty caps is flushed from clients.
> 
> We need to acquire the xlock for the parent inode.
> 
> And also we could voluntarily drop Xx/Lx/Ax caps for some requests to avoid revocation msgs.

So, process_request_cap_release() sends revoke message to a client and starts processing the client request immediately. I wonder why its done that way since ideally it should wait for the revoke ack as its done in path_traverse like your change to acquire parent dir xlock(filelock).

File: src/mds/MDCache.cc (hunk 2 above)
Comment (lxbsz) [reply src/mds/MDCache.cc:2145]: Sure.

File: src/mds/MDCache.cc (hunk 1 above)
Comment (lxbsz) [reply src/mds/MDCache.cc:2158]: sure.

**Comment by vshankar** [comment]
> > If the 'Fx' caps is issued the client will cache the dirty metadatas, while if a write request, mkdir request for example, comes it will try to update the parents' mtime without trigger to revoke the corresponding caps. This will cause the mtime will be overwrote later after the dirty caps is flushed from clients.
> > We need to acquire the xlock for the parent inode.
> > And also we could voluntarily drop Xx/Lx/Ax caps for some requests to avoid revocation msgs.
> 
> So, process_request_cap_release() sends revoke message to a client and starts processing the client request immediately. I wonder why its done that way since ideally it should wait for the revoke ack as its done in path_traverse like your change to acquire parent dir xlock(filelock).

Circling back on this: adding xlock(filelock) in MDCache::path_traverse is correct, however, it could be the case that the revoke ack from the client (from process_request_cap_release()) is processed after revoking Fx caps for the parent directory. Would the mtime be overwritten in this case too? @lxbsz

**Comment by lxbsz** [comment]
> > > If the 'Fx' caps is issued the client will cache the dirty metadatas, while if a write request, mkdir request for example, comes it will try to update the parents' mtime without trigger to revoke the corresponding caps. This will cause the mtime will be overwrote later after the dirty caps is flushed from clients.
> > > We need to acquire the xlock for the parent inode.
> > > And also we could voluntarily drop Xx/Lx/Ax caps for some requests to avoid revocation msgs.
> > 
> > 
> > So, process_request_cap_release() sends revoke message to a client and starts processing the client request immediately. I wonder why its done that way since ideally it should wait for the revoke ack as its done in path_traverse like your change to acquire parent dir xlock(filelock).
> 

@vshankar Sorry for late to reply this.

In the client request it will try to `volunteer` to release some caps those could be used by the MDS side, and it will only `volunteer` to release the caps which are `not used/dirty` in client side.

So in the `process_request_cap_release()` it just try to eval the Lockers to help speed up the client requests handling later, and all the necessary revocations should be done and guaranteed by the client request handling code themselves IMO. 

> Circling back on this: adding xlock(filelock) in MDCache::path_traverse is correct, however, it could be the case that the revoke ack from the client (from process_request_cap_release()) is processed after revoking Fx caps for the parent directory. Would the mtime be overwritten in this case too? @lxbsz

That doesn't matter, when the first revocation ack back it will wake up the waiters and then could continue and correctly handle the `mtime`.

**Comment by lxbsz** [comment]
jenkins test make check

**Comment by dparmar18** [comment]
and i think some what server code could also be fixed
https://github.com/ceph/ceph/blob/70d8c5b565c9f2897714844486d456ad78b0d24d/src/mds/Server.h#L496-L507

**Comment by vshankar** [comment]
jenkins test make check

**Comment by vshankar** [comment]
https://pulpito.ceph.com/?branch=wip-vshankar-testing-20231127.102654

(rhel pkg install failures are a bunch, so, this would need a revalidate)

**Comment by vshankar** [comment]
@lxbsz I think this change is causing the MDS to assert during path traversal and acquiring xlock.

```
2023-12-06T14:40:49.685 INFO:tasks.ceph.mds.b.smithi137.stderr:/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos9/DIST/centos9/MACHINE_SIZE/gigantic/release/19.0.0-22-g9488dd55/rpm/el9/BUILD/ceph-19.0.0-22-g9488dd55/src/mds/Locker.cc: 1963: FAILED ceph_assert(lock->get_sm()->can_remote_xlock)
2023-12-06T14:40:49.687 INFO:tasks.ceph.mds.b.smithi137.stderr: ceph version 19.0.0-22-g9488dd55 (9488dd55a18030b93f338d78f72b42610452f71c) reef (dev)
2023-12-06T14:40:49.687 INFO:tasks.ceph.mds.b.smithi137.stderr: 1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x12e) [0x7f3cf856eef8]
2023-12-06T14:40:49.687 INFO:tasks.ceph.mds.b.smithi137.stderr: 2: /usr/lib64/ceph/libceph-common.so.2(+0x16f0b6) [0x7f3cf856f0b6]
2023-12-06T14:40:49.687 INFO:tasks.ceph.mds.b.smithi137.stderr: 3: (Locker::xlock_start(SimpleLock*, boost::intrusive_ptr<MDRequestImpl>&)+0x939) [0x55da81e3b599]
2023-12-06T14:40:49.688 INFO:tasks.ceph.mds.b.smithi137.stderr: 4: (Locker::acquire_locks(boost::intrusive_ptr<MDRequestImpl>&, MutationImpl::LockOpVec&, CInode*, bool)+0x20f5) [0x55da81e30825]
2023-12-06T14:40:49.688 INFO:tasks.ceph.mds.b.smithi137.stderr: 5: (MDCache::path_traverse(boost::intrusive_ptr<MDRequestImpl>&, ContextFactory<MDSContext>&, filepath const&, int, std::vector<CDentry*, std::allocator<CDentry*> >*, CInode**)+0x79b) [0x55da81ddf34b]
2023-12-06T14:40:49.688 INFO:tasks.ceph.mds.b.smithi137.stderr: 6: (Server::rdlock_path_xlock_dentry(boost::intrusive_ptr<MDRequestImpl>&, bool, bool, bool, bool)+0x371) [0x55da81d19681]
2023-12-06T14:40:49.688 INFO:tasks.ceph.mds.b.smithi137.stderr: 7: (Server::handle_client_unlink(boost::intrusive_ptr<MDRequestImpl>&)+0x90) [0x55da81d3b340]
2023-12-06T14:40:49.688 INFO:tasks.ceph.mds.b.smithi137.stderr: 8: (Server::handle_peer_auth_pin_ack(boost::intrusive_ptr<MDRequestImpl>&, boost::intrusive_ptr<MMDSPeerRequest const> const&)+0x81f) [0x55da81d1602f]
2023-12-06T14:40:49.688 INFO:tasks.ceph.mds.b.smithi137.stderr: 9: (Server::handle_peer_request_reply(boost::intrusive_ptr<MMDSPeerRequest const> const&)+0x6bc) [0x55da81d1691c]
2023-12-06T14:40:49.688 INFO:tasks.ceph.mds.b.smithi137.stderr: 10: (Server::dispatch(boost::intrusive_ptr<Message const> const&)+0x97) [0x55da81d025d7]
2023-12-06T14:40:49.688 INFO:tasks.ceph.mds.b.smithi137.stderr: 11: (MDSRank::_dispatch(boost::intrusive_ptr<Message const> const&, bool)+0x4eb) [0x55da81c8657b]
2023-12-06T14:40:49.688 INFO:tasks.ceph.mds.b.smithi137.stderr: 12: (MDSRankDispatcher::ms_dispatch(boost::intrusive_ptr<Message const> const&)+0x5c) [0x55da81c86a6c]
2023-12-06T14:40:49.688 INFO:tasks.ceph.mds.b.smithi137.stderr: 13: (MDSDaemon::ms_dispatch2(boost::intrusive_ptr<Message> const&)+0x195) [0x55da81c70595]
2023-12-06T14:40:49.688 INFO:tasks.ceph.mds.b.smithi137.stderr: 14: (DispatchQueue::entry()+0x53a) [0x7f3cf876040a]
2023-12-06T14:40:49.688 INFO:tasks.ceph.mds.b.smithi137.stderr: 15: /usr/lib64/ceph/libceph-common.so.2(+0x3f24a1) [0x7f3cf87f24a1]
2023-12-06T14:40:49.688 INFO:tasks.ceph.mds.b.smithi137.stderr: 16: /lib64/libc.so.6(+0x9f802) [0x7f3cf7c9f802]
2023-12-06T14:40:49.689 INFO:tasks.ceph.mds.b.smithi137.stderr: 17: /lib64/libc.so.6(+0x3f450) [0x7f3cf7c3f450]
2023-12-06T14:40:49.689 INFO:tasks.ceph.mds.b.smithi137.stderr:*** Caught signal (Aborted) **
```

/a/vshankar-2023-12-06_12:01:15-fs-wip-vshankar-testing-20231127.102654-r1-distro-default-smithi/7480089

**Comment by lxbsz** [comment]
> @lxbsz I think this change is causing the MDS to assert during path traversal and acquiring xlock.
> 
> ```
> 2023-12-06T14:40:49.685 INFO:tasks.ceph.mds.b.smithi137.stderr:/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos9/DIST/centos9/MACHINE_SIZE/gigantic/release/19.0.0-22-g9488dd55/rpm/el9/BUILD/ceph-19.0.0-22-g9488dd55/src/mds/Locker.cc: 1963: FAILED ceph_assert(lock->get_sm()->can_remote_xlock)
> 2023-12-06T14:40:49.687 INFO:tasks.ceph.mds.b.smithi137.stderr: ceph version 19.0.0-22-g9488dd55 (9488dd55a18030b93f338d78f72b42610452f71c) reef (dev)
> 2023-12-06T14:40:49.687 INFO:tasks.ceph.mds.b.smithi137.stderr: 1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x12e) [0x7f3cf856eef8]
> 2023-12-06T14:40:49.687 INFO:tasks.ceph.mds.b.smithi137.stderr: 2: /usr/lib64/ceph/libceph-common.so.2(+0x16f0b6) [0x7f3cf856f0b6]
> 2023-12-06T14:40:49.687 INFO:tasks.ceph.mds.b.smithi137.stderr: 3: (Locker::xlock_start(SimpleLock*, boost::intrusive_ptr<MDRequestImpl>&)+0x939) [0x55da81e3b599]
> 2023-12-06T14:40:49.688 INFO:tasks.ceph.mds.b.smithi137.stderr: 4: (Locker::acquire_locks(boost::intrusive_ptr<MDRequestImpl>&, MutationImpl::LockOpVec&, CInode*, bool)+0x20f5) [0x55da81e30825]
> 2023-12-06T14:40:49.688 INFO:tasks.ceph.mds.b.smithi137.stderr: 5: (MDCache::path_traverse(boost::intrusive_ptr<MDRequestImpl>&, ContextFactory<MDSContext>&, filepath const&, int, std::vector<CDentry*, std::allocator<CDentry*> >*, CInode**)+0x79b) [0x55da81ddf34b]
> 2023-12-06T14:40:49.688 INFO:tasks.ceph.mds.b.smithi137.stderr: 6: (Server::rdlock_path_xlock_dentry(boost::intrusive_ptr<MDRequestImpl>&, bool, bool, bool, bool)+0x371) [0x55da81d19681]
> 2023-12-06T14:40:49.688 INFO:tasks.ceph.mds.b.smithi137.stderr: 7: (Server::handle_client_unlink(boost::intrusive_ptr<MDRequestImpl>&)+0x90) [0x55da81d3b340]
> 2023-12-06T14:40:49.688 INFO:tasks.ceph.mds.b.smithi137.stderr: 8: (Server::handle_peer_auth_pin_ack(boost::intrusive_ptr<MDRequestImpl>&, boost::intrusive_ptr<MMDSPeerRequest const> const&)+0x81f) [0x55da81d1602f]
> 2023-12-06T14:40:49.688 INFO:tasks.ceph.mds.b.smithi137.stderr: 9: (Server::handle_peer_request_reply(boost::intrusive_ptr<MMDSPeerRequest const> const&)+0x6bc) [0x55da81d1691c]
> 2023-12-06T14:40:49.688 INFO:tasks.ceph.mds.b.smithi137.stderr: 10: (Server::dispatch(boost::intrusive_ptr<Message const> const&)+0x97) [0x55da81d025d7]
> 2023-12-06T14:40:49.688 INFO:tasks.ceph.mds.b.smithi137.stderr: 11: (MDSRank::_dispatch(boost::intrusive_ptr<Message const> const&, bool)+0x4eb) [0x55da81c8657b]
> 2023-12-06T14:40:49.688 INFO:tasks.ceph.mds.b.smithi137.stderr: 12: (MDSRankDispatcher::ms_dispatch(boost::intrusive_ptr<Message const> const&)+0x5c) [0x55da81c86a6c]
> 2023-12-06T14:40:49.688 INFO:tasks.ceph.mds.b.smithi137.stderr: 13: (MDSDaemon::ms_dispatch2(boost::intrusive_ptr<Message> const&)+0x195) [0x55da81c70595]
> 2023-12-06T14:40:49.688 INFO:tasks.ceph.mds.b.smithi137.stderr: 14: (DispatchQueue::entry()+0x53a) [0x7f3cf876040a]
> 2023-12-06T14:40:49.688 INFO:tasks.ceph.mds.b.smithi137.stderr: 15: /usr/lib64/ceph/libceph-common.so.2(+0x3f24a1) [0x7f3cf87f24a1]
> 2023-12-06T14:40:49.688 INFO:tasks.ceph.mds.b.smithi137.stderr: 16: /lib64/libc.so.6(+0x9f802) [0x7f3cf7c9f802]
> 2023-12-06T14:40:49.689 INFO:tasks.ceph.mds.b.smithi137.stderr: 17: /lib64/libc.so.6(+0x3f450) [0x7f3cf7c3f450]
> 2023-12-06T14:40:49.689 INFO:tasks.ceph.mds.b.smithi137.stderr:*** Caught signal (Aborted) **
> ```
> 
> /a/vshankar-2023-12-06_12:01:15-fs-wip-vshankar-testing-20231127.102654-r1-distro-default-smithi/7480089

@vshankar I found the root cause for this failure. For example for '/mydir/fileA`, when deleting the 'fileA' the clients will send the `unlink` request to the `fileA` dentry's `auth MDS`.

But for `mydir/` directory the corresponding inode's `auth MDS` won't be always be the same with `fileA`. So when we try to get the `xlock` of `mydir/` it will trigger the above ceph assert.

Currently for the 'sm_filelock' it won't allow to do the remote xlock, please see `can_remote_xlock`:

```
126 const struct sm_t sm_filelock = {
127         .states = filelock,
128         .allowed_ever_auth = (CEPH_CAP_GSHARED |
129                               CEPH_CAP_GEXCL |
130                               CEPH_CAP_GCACHE |
131                               CEPH_CAP_GRD |
132                               CEPH_CAP_GWR |
133                               CEPH_CAP_GWREXTEND |
134                               CEPH_CAP_GBUFFER | 
135                               CEPH_CAP_GLAZYIO),
136         .allowed_ever_replica = (CEPH_CAP_GSHARED |
137                                  CEPH_CAP_GCACHE |
138                                  CEPH_CAP_GRD | 
139                                  CEPH_CAP_GLAZYIO),
140         .careful = (CEPH_CAP_GSHARED | 
141                     CEPH_CAP_GEXCL | 
142                     CEPH_CAP_GCACHE |
143                     CEPH_CAP_GBUFFER),
144         .can_remote_xlock = 0,
145 };
```

By going through the code I didn't find any reason why we couldn't. Currently only the `sm_simplelock` allows it. Any thought or caution about enabling it ? 

Locally I just enabled it and now still running the tests, till now everything worked well.

@batrick @gregsfortytwo Any thought ? Thanks.

**Comment by mchangir** [comment]
Teuthology Job Status

Job: snap-schedule
URL: https://pulpito.ceph.com/mchangir-2024-03-22_09:41:51-fs-wip-mchangir-testing-main-20240318.032620-testing-default-smithi/
Status: 11 jobs failed with PG_AVAILABILITY issues, whereas they actual tests passed as seen below

```
mchangir:logs$ for log in $(find /teuthology/mchangir-2024-03-22_09:41:51-fs-wip-mchangir-testing-main-20240318.032620-testing-default-smithi -mindepth 2 -maxdepth 2 -name 'teuthology.log'); do grep -H 'cephfs_test_runner:OK' $log; done
/teuthology/mchangir-2024-03-22_09:41:51-fs-wip-mchangir-testing-main-20240318.032620-testing-default-smithi/7616249/teuthology.log:2024-03-22T12:09:40.070 INFO:tasks.cephfs_test_runner:OK
/teuthology/mchangir-2024-03-22_09:41:51-fs-wip-mchangir-testing-main-20240318.032620-testing-default-smithi/7616252/teuthology.log:2024-03-22T12:11:43.760 INFO:tasks.cephfs_test_runner:OK
/teuthology/mchangir-2024-03-22_09:41:51-fs-wip-mchangir-testing-main-20240318.032620-testing-default-smithi/7616255/teuthology.log:2024-03-22T12:10:44.872 INFO:tasks.cephfs_test_runner:OK
/teuthology/mchangir-2024-03-22_09:41:51-fs-wip-mchangir-testing-main-20240318.032620-testing-default-smithi/7616251/teuthology.log:2024-03-22T12:20:18.371 INFO:tasks.cephfs_test_runner:OK
/teuthology/mchangir-2024-03-22_09:41:51-fs-wip-mchangir-testing-main-20240318.032620-testing-default-smithi/7616258/teuthology.log:2024-03-22T12:15:45.875 INFO:tasks.cephfs_test_runner:OK
/teuthology/mchangir-2024-03-22_09:41:51-fs-wip-mchangir-testing-main-20240318.032620-testing-default-smithi/7616254/teuthology.log:2024-03-22T12:20:04.651 INFO:tasks.cephfs_test_runner:OK
/teuthology/mchangir-2024-03-22_09:41:51-fs-wip-mchangir-testing-main-20240318.032620-testing-default-smithi/7616248/teuthology.log:2024-03-22T12:22:11.607 INFO:tasks.cephfs_test_runner:OK
/teuthology/mchangir-2024-03-22_09:41:51-fs-wip-mchangir-testing-main-20240318.032620-testing-default-smithi/7616250/teuthology.log:2024-03-22T12:23:09.965 INFO:tasks.cephfs_test_runner:OK
/teuthology/mchangir-2024-03-22_09:41:51-fs-wip-mchangir-testing-main-20240318.032620-testing-default-smithi/7616256/teuthology.log:2024-03-22T12:22:24.853 INFO:tasks.cephfs_test_runner:OK
/teuthology/mchangir-2024-03-22_09:41:51-fs-wip-mchangir-testing-main-20240318.032620-testing-default-smithi/7616257/teuthology.log:2024-03-22T12:20:46.343 INFO:tasks.cephfs_test_runner:OK
/teuthology/mchangir-2024-03-22_09:41:51-fs-wip-mchangir-testing-main-20240318.032620-testing-default-smithi/7616253/teuthology.log:2024-03-22T12:22:12.682 INFO:tasks.cephfs_test_runner:OK
```

Job: fs:upgrade
URL: https://pulpito.ceph.com/mchangir-2024-03-22_09:46:06-fs:upgrade-wip-mchangir-testing-main-20240318.032620-testing-default-smithi/
Status: 5 of 13 failed

all are known issues


Job: fs:libcephfs
URL: https://pulpito.ceph.com/mchangir-2024-03-22_09:48:09-fs:libcephfs-wip-mchangir-testing-main-20240318.032620-testing-default-smithi/
Status: 2 fail, 2 pass

all are known issues


Job: snap_schedule
URL: https://pulpito.ceph.com/mchangir-2024-03-22_09:40:19-fs-wip-mchangir-testing-main-20240318.032620-testing-default-smithi/
Status: 12 of 12 passed

Wiki: https://tracker.ceph.com/projects/cephfs/wiki/Main#2024-03-25

**Comment by lxbsz** [comment]
> Teuthology Job Status
> 
> Job: snap-schedule URL: https://pulpito.ceph.com/mchangir-2024-03-22_09:41:51-fs-wip-mchangir-testing-main-20240318.032620-testing-default-smithi/ Status: 11 jobs failed with PG_AVAILABILITY issues, whereas they actual tests passed as seen below
> 

@mchangir Sorry for missing this message.

BTW, do you mean the `PG_AVAILABILITY` issue is related to this ?

File: src/mds/MDCache.h [hunk 3]
```diff
@@ -118,18 +118,19 @@ enum {
...
+static const int MDS_TRAVERSE_WANT_AUTH          = (1 << 3);
+static const int MDS_TRAVERSE_RDLOCK_SNAP        = (1 << 4);
+static const int MDS_TRAVERSE_RDLOCK_SNAP2       = (1 << 5);
+static const int MDS_TRAVERSE_WANT_DIRLAYOUT     = (1 << 6);
+static const int MDS_TRAVERSE_RDLOCK_PATH        = (1 << 7);
+static const int MDS_TRAVERSE_XLOCK_DENTRY       = (1 << 8);
+static const int MDS_TRAVERSE_RDLOCK_AUTHLOCK    = (1 << 9);
+static const int MDS_TRAVERSE_CHECK_LOCKCACHE    = (1 << 10);
+static const int MDS_TRAVERSE_WANT_INODE         = (1 << 11);
+static const int MDS_TRAVERSE_XLOCK_PARENT_INODE = (1 << 12);
```
Comment (batrick) [inline src/mds/MDCache.h:133]: I don't think it's worth trying to line these up. It messes up the blame history every time. Let's just keep one space on both sides of the `=`.

File: src/mds/MDCache.cc [hunk 4]
```diff
@@ -8424,16 +8424,25 @@ int MDCache::path_traverse(MDRequestRef& mdr, MDSContextFactory& cf,
 	    depth == path.depth() - 1) {
 	  ceph_assert(dn->is_auth());
 	  if (depth > 0 || !mdr->lock_cache) {
-	    lov.add_wrlock(&cur->filelock);
+            if (xlock_dentry_parent) {
+              lov.add_xlock(&cur->filelock);
```
Comment (batrick) [inline src/mds/MDCache.cc:8428]: This won't be appropriate I think. The `filelock` is deliberately configured to allow multiple writers (yes, even and especially for directory inodes) otherwise we will serialize every file create. (Please test this to confirm.)

What appears to be missing is that a wrlock on the `filelock` is not correctly revoking (or waiting for the revoke of) `Fx` on the directory inode. No?

**Review by batrick — CHANGES_REQUESTED** [review]
The test case was simple, let's add it?

File: src/mds/MDCache.h (hunk 3 above)
Comment (lxbsz) [reply src/mds/MDCache.h:133]: Sure. Will fix it.

File: src/mds/MDCache.cc (hunk 4 above)
Comment (lxbsz) [reply src/mds/MDCache.cc:8428]: > This won't be appropriate I think. The `filelock` is deliberately configured to allow multiple writers (yes, even and especially for directory inodes) otherwise we will serialize every file create. (Please test this to confirm.)
> 

Yeah, this really will serialize the creates.

> What appears to be missing is that a wrlock on the `filelock` is not correctly revoking (or waiting for the revoke of) `Fx` on the directory inode. No?

Yeah, correct. Maybe we can fix it by revoking the `Fx` when acquiring the `wrlock` ?

**Comment by lxbsz** [comment]
> Currently only the `sm_simplelock` allows it

Sure.

File: src/mds/MDCache.cc (hunk 4 above)
Comment (lxbsz) [reply src/mds/MDCache.cc]: Tested the new patch and it worked well.

File: src/mds/SimpleLock.h [hunk 5]
```diff
@@ -359,13 +359,23 @@ class SimpleLock {
     return get_sm()->states[state].can_wrlock == ANY ||
       (get_sm()->states[state].can_wrlock == AUTH && parent->is_auth()) ||
       (get_sm()->states[state].can_wrlock == XCL && client >= 0 && (get_xlock_by_client() == client ||
-								    get_excl_client() == client));
+                                                                    (get_excl_client() == client &&
+                                                                     !(parent->is_dir() &&
+                                                                       (type->type == CEPH_LOCK_IFILE ||
+                                                                        type->type == CEPH_LOCK_IAUTH ||
+                                                                        type->type == CEPH_LOCK_ILINK ||
+                                                                        type->type == CEPH_LOCK_IXATTR)))));
```
Comment (batrick) [inline src/mds/SimpleLock.h:367]: Is this problem actually unique to directories? Don't we want to revoke `Fx` even for regular files?

File: src/mds/locks.c [hunk 6]
```diff
@@ -142,7 +142,7 @@ const struct sm_t sm_filelock = {
 		    CEPH_CAP_GEXCL | 
 		    CEPH_CAP_GCACHE |
 		    CEPH_CAP_GBUFFER),
-	.can_remote_xlock = 0,
+	.can_remote_xlock = 1,
```
Comment (batrick) [inline src/mds/locks.c:145]: revert

File: src/test/libcephfs/test.cc [hunk 7]
```diff
@@ -2263,6 +2263,30 @@ TEST(LibCephFS, TestFutimens) {
...
+  struct ceph_statx stx1, stx2;
+  ASSERT_EQ(ceph_statx(cmount, c_dir, &stx1, CEPH_STATX_MTIME, 0), 0);
+  sleep(1);
+  sprintf(c_subdir, "%s/%s", c_dir, "mtime_subdir");
+  ASSERT_EQ(0, ceph_mkdir(cmount, c_subdir, 0777));
+
+  ASSERT_EQ(ceph_statx(cmount, c_dir, &stx2, CEPH_STATX_MTIME, 0), 0);
+  ASSERT_NE(utime_t(stx1.stx_mtime), utime_t(stx2.stx_mtime));
+
+  ceph_shutdown(cmount);
```
Comment (batrick) [inline src/test/libcephfs/test.cc:2287]: :+1:

File: src/mds/MDSCacheObject.h [hunk 8]
```diff
@@ -119,6 +120,7 @@ class MDSCacheObject {
   void state_set(unsigned mask) { state |= mask; }
   void state_reset(unsigned s) { state = s; }
 
+  bool is_dir() const { return state_test(STATE_DIR); }
```
Comment (batrick) [inline src/mds/MDSCacheObject.h:123]: better would be a `virtual bool is_dir() const` that is overriden here (and defined in `MDSCacheObject`.

File: src/client/Client.cc [hunk 9]
```diff
@@ -2160,16 +2160,26 @@ int Client::encode_inode_release(Inode *in, MetaRequest *req,
...
+	drop &= ~CEPH_CAP_LINK_SHARED;
+      if (unless & CEPH_CAP_XATTR_EXCL)
+	drop &= ~CEPH_CAP_XATTR_SHARED;
+      if (unless & CEPH_CAP_FILE_EXCL)
+	drop &= ~CEPH_CAP_FILE_SHARED;
+    }
+
+    if (force || (drop & cap.issued)) {
+      if (drop & cap.issued) {
+        ldout(cct, 25) << "dropping caps " << ccap_string(drop) << dendl;
```
Comment (batrick) [inline src/client/Client.cc:2178]:
```suggestion
        ldout(cct, 20) << "dropping caps " << ccap_string(drop) << dendl;
```

**Review by batrick — CHANGES_REQUESTED** [review]

File: src/mds/SimpleLock.h (hunk 5 above)
Comment (lxbsz) [reply src/mds/SimpleLock.h:367]: Only for parent directory will have this issue.

File: src/mds/MDSCacheObject.h (hunk 8 above)
Comment (lxbsz) [reply src/mds/MDSCacheObject.h:123]: Sure.

File: src/client/Client.cc (hunk 9 above)
Comment (lxbsz) [reply src/client/Client.cc:2178]: Will fix it.

File: src/mds/locks.c (hunk 6 above)
Comment (lxbsz) [reply src/mds/locks.c]: Done.

File: src/mds/CDir.h [hunk 10]
```diff
@@ -497,6 +497,8 @@ class CDir : public MDSCacheObject, public Counter<CDir> {
 
   void mark_complete();
 
+  bool is_dir() const { return true; }
```
Comment (batrick) [inline src/mds/CDir.h:500]:
```suggestion
  bool is_dir() const override { return true; }
```

**Review by batrick — APPROVED** [review]

File: src/mds/CDir.h (hunk 10 above)
Comment (lxbsz) [reply src/mds/CDir.h]: Done.

**Comment by vshankar** [comment]
@mchangir pinging on this since I see your testing tag.

**Comment by vshankar** [comment]
> @mchangir pinging on this since I see your testing tag.

ping @mchangir ?
